UPGRADING.md

```diff
@@ -0,0 +1,516 @@
+
+## Breaking Changes
+
+### `ensureReady` is no longer available
+
+there is no need to use `ensureReady` method and it's removed from After.js.
+
+From:
+
+```jsx
+// client.js
+
+import React from 'react';
+import { hydrate } from 'react-dom';
+import { BrowserRouter } from 'react-router-dom';
+import { ensureReady, After } from '@jaredpalmer/after';
+import './client.css';
+import routes from './routes';
+
+ensureReady(routes).then(data =>
+  hydrate(
+    <BrowserRouter>
+      <After data={data} routes={routes} />
+    </BrowserRouter>,
+    document.getElementById('root')
+  )
+);
+
+if (module.hot) {
+  module.hot.accept();
+}
+```
+
+To:
+
+```jsx
+// client.js
+
+import React from 'react';
+import { hydrate } from 'react-dom';
+import { BrowserRouter } from 'react-router-dom';
+import { After } from '@jaredpalmer/after';
+import './client.css';
+import routes from './routes';
+
+hydrate(
+  <BrowserRouter>
+    <After routes={routes} />
+  </BrowserRouter>,
+  document.getElementById('root')
+);
+
+if (module.hot) {
+  module.hot.accept();
+}
+```
+
+You can access server app state from the window (`window.__SERVER_APP_STATE__`), you can modify `SERVER_APP_STATE` and pass it to `<After />` component.
+
+```jsx
+<After data={modifyData(window.__SERVER_APP_STATE__)} routes={routes} />
+```
+
+`After` component will pick `window.__SERVER_APP_STATE__` if data prop not passed.
+
+```jsx
+function After({ routes, data = window.__SERVER_APP_STATE__ }) {
+  // ...
+}
+```
+
+### `routes.js` config changed
+
+To send assets (CSS and JS files) from the initial server response, we need to change how we define our routes.
+
+From:
+
+```jsx
+// routes.js
+
+import Home from './Home';
+import { asyncComponent } from '@jaredpalmer/after';
+
+export default [
+  {
+    path: '/',
+    exact: true,
+    component: Home,
+  },
+  {
+    path: '/about',
+    exact: true,
+    component: asyncComponent({
+      loader: () => import('./About'),
+    }),
+  },
+  {
+    path: '/contact-us',
+    exact: true,
+    component: asyncComponent({
+      loader: () => import('./Contact'),
+    }),
+  },
+];
+```
+
+To:
+
+```jsx
+// routes.js
+
+import Home from './Home';
+import { asyncComponent } from '@jaredpalmer/after';
+
+export default [
+  {
+    path: '/',
+    exact: true,
+    component: Home,
+  },
+  {
+    path: '/about',
+    exact: true,
+    component: asyncComponent({
+      loader: () => import(/* webpackChunkName: "whatever" */ './About'),
+      chunkName: 'whatever',
+    }),
+  },
+  {
+    path: '/contact-us',
+    exact: true,
+    component: asyncComponent({
+      loader: () => import(/* webpackChunkName: "ContactUs" */ './Contact'),
+      chunkName: 'ContactUs',
+    }),
+  },
+];
+```
+
+> This is too hard and complicated (and there are more things that we have to take care about) so we made a babel plugin to do this automatically because we care about you (DX).
```

Comment:
Part of the project philosophy is to _not need_ to rely on Babel. What other things need to be done? We should show the user in its entirety.

Reply:
I updated the UPGRADING.md as shown below

there is one more thing that we have to take care about. if you use the same component in different routes, `webpackChunkName` and `chunkName` values must be the same in all of them.

```jsx
[
  {
    path: '/shop/:filter([A-Za-z-]+)', // 👈 different routes
    exact: true,
    component: asyncComponent({
      loader: () => import(`pages/Shop`), // 👈 same components
    }),
  },
  {
    path: '/shop/:page([0-9]+)?', // 👈 different routes
    exact: true,
    component: asyncComponent({
      loader: () => import(`pages/Shop`), // 👈 same components
    }),
  },
];
```

✅ The right way to handle the above situation:

```jsx
[
  {
    path: '/shop/:filter([A-Za-z-]+)',
    exact: true,
    component: asyncComponent({
      loader: () => import(/* webpackChunkName: "pages-Shop" */ `pages/Shop`),
      chunkName: 'pages-Shop', // 👈  names are identical 👆
    }),
  },
  {
    path: '/shop/:page([0-9]+)?',
    exact: true,
    component: asyncComponent({
      loader: () => import(/* webpackChunkName: "pages-Shop" */ `pages/Shop`),
      chunkName: 'pages-Shop', // 👈  names are identical 👆 and they match with the previous route
    }),
  },
];
```